### PR TITLE
Make NuGet service index publicly accessible

### LINKED ISSUE #20717: NuGet push returns 401 (Unauthorized) when using `--api-key`
### Description

I created an organization with my main account. I can use my account to push nuget packages in to the repository for my organization. Then I generated an API token using the GUI, saved that to my password manager. I've been having trouble using that token. 

When I run this command:

    dotnet nuget push --source gitea {package_file} --api-key {gitea_token}

The response is:

> error: Response status code does not indicate success: 401 (Unauthorized).

### Gitea Version

Gitea Version: 1.17.0

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

2022/08/04 14:59:45 ...kages/nuget/nuget.go:25:apiError() [E] [62ec4151-12] context canceled

### Screenshots

_No response_

### Git Version

git version 2.31.1

### Operating System

Operating System: Red Hat Enterprise Linux 8.6 (Ootpa)        CPE OS Name: cpe:/o:redhat:enterprise_linux:8::baseos             Kernel: Linux 4.18.0-372.16.1.el8_6.x86_64

### How are you running Gitea?

Installed from binary, running as a service.

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by KN4CK3R**
Addition to #20734
Fixes #20717

The `/index.json` endpoint needs to be accessible even if the registry is private. The NuGet client uses this endpoint without authentification.

The old fix only works if the NuGet cli is used with `--source <name>` but not with `--source <url>/index.json`.

### CODE REVIEW

**Comment by silverwind** [comment]
What is the downside? Unauthenticated clients can enumerate NuGet package names?

**Comment by KN4CK3R** [comment]
> So, unauthenticated clients can enumerate NuGet package names? Anything else exposed?

No, that would be really bad. The service index exposes the capabilities and endpoints of the NuGet registry. All other requests are authenticated.
The response looks like this:
```json
{
  "version": "3.0.0",
  "resources": [
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/query",
      "@type": "SearchQueryService"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/query",
      "@type": "SearchQueryService/3.0.0-beta"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/query",
      "@type": "SearchQueryService/3.0.0-rc"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/registration",
      "@type": "RegistrationsBaseUrl"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/registration",
      "@type": "RegistrationsBaseUrl/3.0.0-beta"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/registration",
      "@type": "RegistrationsBaseUrl/3.0.0-rc"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/package",
      "@type": "PackageBaseAddress/3.0.0"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget",
      "@type": "PackagePublish/2.0.0"
    },
    {
      "@id": "http://host.docker.internal:3000/api/packages/KN4CK3R/nuget/symbolpackage",
      "@type": "SymbolPackagePublish/4.9.0"
    }
  ]
}
```

**Review by silverwind — APPROVED** [review]
~~A test would be nice.~~

**Comment by KN4CK3R** [comment]
I already adjusted an existing test for a normal and a private user.

File: routers/api/packages/api.go [hunk 1]
```diff
@@ -169,33 +169,35 @@ func Routes(ctx gocontext.Context) *web.Route {
...
-				r.Get("/index.json", nuget.EnumeratePackageVersions)
-				r.Get("/{version}/{filename}", nuget.DownloadPackageFile)
-			})
 			r.Group("", func() {
-				r.Put("/", nuget.UploadPackage)
-				r.Put("/symbolpackage", nuget.UploadSymbolPackage)
-				r.Delete("/{id}/{version}", nuget.DeletePackage)
-			}, reqPackageAccess(perm.AccessModeWrite))
-			r.Get("/symbols/{filename}/{guid:[0-9a-f]{32}}FFFFFFFF/{filename2}", nuget.DownloadSymbolFile)
+				r.Get("/query", nuget.SearchService)
```
Comment (wxiaoguang) [inline routers/api/packages/api.go:185]: LGTM, while I think it needs more comments about why there is no permission check for it.

And, maybe someone will worry about "attackers can detect private user names" again, while I think it's acceptable and that's cost of using NuGet package registry.

Comment (KN4CK3R) [reply routers/api/packages/api.go:185]: > And, maybe someone will worry about "attackers can detect private user names" again, while I think it's acceptable and that's cost of using NuGet package registry.

You can detect that with the `/users/{user}` api too, so I see no difference.

Comment (wxiaoguang) [reply routers/api/packages/api.go:185]: I agree, so LGTM 😁

**Review by wxiaoguang — APPROVED** [review]

File: routers/api/packages/api.go (hunk 1 above)
Comment (KN4CK3R) [reply routers/api/packages/api.go:185]: I think that's a definition problem of private. Private like "there is a locked door I know something is behind" or private like "which door?".

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**